Truncate

AGaliciaMartinez opened the pull request:
This PR includes a new method `truncate`  together with some small fixed for the tensor train class.

A few notes on this method. Tests currently ensure that the output makes sense as a tensor train. They also ensure that the bond dimension o the output has been truncated. I have not been able to design a test that numerically checks for the accuracy of the truncation step though. The issue here is that I could not think of an easy analytical example to compare to. Also the "error" returned does not accurately reproduce (and it should not) the expected error obtained from the norm of `truncated_state - state`. Which makes it difficult to relate those two values in a test.

The tensor train is left in a right normalized state. This means that the nodes are now unitary matrices (when shaped correctly). However, I am worried that a truncation with only a right normalization may not be enough. I am worried in particular for the case of using max_trunc_error. The point is that after truncating it for the first time, the all nodes are unitary and hence have a degenerate set of singular values equal to one. I am wondering how another truncation will affect the normalized state. The point is that if you have to get rid of a few singular values but all are the same... how do you know which one is bet to get rid of? I guess this should never happen in practice as you should only truncate the tensor train once the bond dimension has increased and hence the singular values will probably not be degenerated any-more. I guess this will come out again once matmul is included. For the moment lets just merge as it is and we will improve it later if needed.


Changelog:
- New method truncate added: it performs from left to right and svd truncation in the nodes of a tensor train. 
- Copy method fixed: previously it returned a Network not a tensortrain.

AGaliciaMartinez committed:
Added tests and changed _fast_from... method.

- Tests only ensure that the output is a valid tensor train. They do not
check for numerical accuracy.
- I also changed the _fast_from method to work with TT. In particular,
the change is made such that copy works with tt.

AGaliciaMartinez committed:
Added tests for copy in FiniteTT.

AGaliciaMartinez committed:
Merge branch 'main' into truncate

AGaliciaMartinez committed:
Changed name of test for copy tensor trains.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez committed:
Fixed tests.

AGaliciaMartinez committed:
Removed relative keyword as it is not being used.

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -191,7 +191,7 @@ def copy(self):
         out_edges = [edge_dict[e] for e in self.out_edges]
 
-        return Network._fast_constructor(out_edges, in_edges, nodes)
+        return self._fast_constructor(out_edges, in_edges, nodes)
```
:+1:

hodgestar commented on src/qutip_tensornetwork/core/data/tensor_train/tensor_train.py:
```diff
@@ -234,6 +234,119 @@ def _to_tt_format(self):
+        Notes
+        -----
+        The truncation method consists on performing a svd decomposition from
+        left to right and truncating each nodes svd by the given parameters.
```
```suggestion
        The truncation method consists of performing an SVD decomposition from
        left to right and truncating each node's list of singular values as specified
        (see the parameter descriptions below).
```

hodgestar commented on src/qutip_tensornetwork/core/data/tensor_train/tensor_train.py:
```diff
@@ -234,6 +234,119 @@ def _to_tt_format(self):
+
+        if not isinstance(max_truncation_err, list):
+            max_truncation_err = [max_truncation_err] * len(self.bond_edges)
+
```
We should check that `len(bond_dimension) == len(self.bond_edges)` and `len(max_truncation_err) == len(self.bond_edges)` and raise a ValueError if they are not (since otherwise `zip(self.bond_edges, ...)` will silently drop any extra elements and create a mess.

hodgestar commented on src/qutip_tensornetwork/core/data/tensor_train/tensor_train.py:
```diff
@@ -234,6 +234,119 @@ def _to_tt_format(self):
+            same bond dimension for all bonds. None is understood as
+            "infinite" bond dimension and hence no truncation is done.
+
+        max_truncation_err: List of float or float
```
We should:
- method that `max_truncation_err` can also be `None` (what that means)
- explain a bit more how the bond_dimension and max_truncation_err interact (i.e. bond_dimension takes precedence)

hodgestar commented on src/qutip_tensornetwork/core/data/tensor_train/tensor_train.py:
```diff
@@ -234,6 +234,119 @@ def _to_tt_format(self):
+                max_singular_values=dim,
+                max_truncation_err=err,
+            )
+            total_error += error.tolist()
```
I wonder how useful it is to return a flat list here? If we return a flat list, then one gets `[err1, err2, err3, ...]` but one has no idea which edge the singular values came from because one doesn't know how many were truncated each time. Perhaps it would be better to do `total_error.append(error.tolist())`? Or `total_error.append(error)` if the error object is useful to have (maybe it's a tensor node)?

hodgestar commented on src/qutip_tensornetwork/core/data/tensor_train/tensor_train.py:
```diff
@@ -246,14 +359,14 @@ def _check_shape(nodes):
         if len(node.shape) != 1 + len(nodes[0].shape):
             raise ValueError(
-                " the shape of the {i}-th node is not correct. It"
+                f" the shape of the {i}-th node is not correct. It"
```
```suggestion
                f"The shape of the {i}-th node is not correct. It"
```

hodgestar commented on tests/core/data/test_tensor_train.py:
```diff
@@ -32,6 +32,23 @@ def assert_bond_edges_name(tt):
+    """Create a random mpo with n sites d dimension per site and
+    bond_dimesnion."""
+    if n > 1:
+        list_tensors = [np.random.random((d, d, bond_dimension)) - 1 / 2]
```
I wonder if we should add some imaginary components here to make these tensors complex so that we can test that conjugation happens correctly?

hodgestar commented on tests/core/data/test_tensor_train.py:
```diff
@@ -245,3 +262,106 @@ def test_from_2d_array(shape, expected_dims):
+        error = mpo.truncate()
+        assert len(error) == 0
+        # The default values will truncate the final tensor train as the first
+        # few bond dimensions are lager than they need to be.
```
```suggestion
        # few bond dimensions are larger than they need to be.
```

hodgestar reviewed (changes requested):
The implementation is really readable -- thank you for the great ASCII pictures in the code. Those were very helpful.

Maybe we could check the mathematical correctness by contracting the whole network at the end and comparing the large singular values of the original and truncated networks?

Regarding the direction of truncation: To me it seems more "natural" to contract the node with the singular values to the left when proceeding to the right as we currently do. That leaves the singular values more distributed within the train, rather than all in the right most node as currently happens. I understand that doesn't leave the network in a canonical form though. I leave the choice of whether to keep this as is for now up to you. It sounded from our conversation this morning that the currently implementation is the right one for the examples you've tried so far.

In the long run I guess it would make sense to support both moving left or right through the train and contracting the singular values either left or right, so that users could pick whichever option made the most sense to them (but I definitely don't know how people use these truncations in practice yet, so please take this idea with a pitch of salt).

AGaliciaMartinez committed:
Update src/qutip_tensornetwork/core/data/tensor_train/tensor_train.py

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez committed:
Apply suggestions from code review

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez committed:
Added a fix for truncate.

Since we do not ensure that nodes in edges are sorted, we can not rely
on node1 and node2 to get the nodes.

AGaliciaMartinez committed:
Changed behaviour to contract the singular values to the left node instead of right node.

AGaliciaMartinez committed:
Changed total_error for truncated values.

It now returns a list of list with the truncated singular values for
each node.

AGaliciaMartinez committed:
Chnged total_error to truncated_values.

`truncate` now returns a list of lists with the truncated singular
values.

AGaliciaMartinez committed:
Merge branch 'truncate' of github.com:AGaliciaMartinez/qutip-tensornetwork into truncate

AGaliciaMartinez committed:
Changed random node generation to include complex values.

Previously, random nodes contained only real values. Now the contain
scalar values.

AGaliciaMartinez committed:
Improved `truncate` docstring.

AGaliciaMartinez committed:
Added len check for truncate arguments.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez commented:
I ended up changing the algorithm to contract the singular values to the left instead of to the right. I think it is now easier to create a test for numerical accuracy so I will do so tomorrow. After that it should be ready to be merged.

AGaliciaMartinez committed:
Added test to check numerical accuracy of the truncation step.

AGaliciaMartinez committed:
Changed order of edges for the svd process.

This should not affect the value of singular values nor its truncation
but it does make it easier to test the function.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez commented:
Added on more tests that check for numerical correctness of the algorithm. It does so by using an mpo that has nodes that are diagonal and hence it is easier to infer the resulting tensor train from the truncation.

Let me know if you are happy with this @hodgestar and I will merge it.
